crates/store/src/state.rs

```diff
@@ -494,6 +496,145 @@ impl State {
+            .db
+            .select_note_inclusion_proofs(unauthenticated_note_ids)
+            .await
+            .map_err(GetBatchInputsError::SelectNoteInclusionProofError)?;
+
+        // The set of blocks that the notes are included in.
+        let note_blocks = note_proofs.values().map(|proof| proof.location().block_num());
+
+        // Collect all blocks we need to query without duplicates, which is:
+        // - all blocks for which we need to prove note inclusion.
+        // - all blocks referenced by transactions in the batch.
+        let mut blocks = tx_reference_blocks;
+        blocks.extend(note_blocks);
+
+        // Grab the block merkle paths from the inner state.
+        //
+        // NOTE: Scoped block to automatically drop the mutex guard asap.
+        //
+        // We also avoid accessing the db in the block as this would delay
+        // dropping the guard.
+        let (batch_reference_block, partial_mmr) = {
+            let state = self.inner.read().await;
+            let latest_block_num = state.latest_block_num();
+
+            let highest_block_num =
+                *blocks.last().expect("we should have checked for empty block references");
+            if highest_block_num > latest_block_num {
+                return Err(GetBatchInputsError::TransactionBlockReferenceNewerThanLatestBlock {
+                    highest_block_num,
+                    latest_block_num,
+                });
+            }
+
+            // Remove the latest block from the to-be-tracked blocks as it will be the reference
+            // block for the batch itself and thus added to the MMR within the batch kernel, so
+            // there is no need to prove its inclusion.
+            blocks.remove(&latest_block_num);
+
+            // Using latest block as the target forest means we take the state of the MMR one before
+            // the latest block. This is because the latest block will be used as the reference
+            // block of the batch and will be added to the MMR by the batch kernel.
+            let target_forest = latest_block_num.as_usize();
+            let peaks = state
+                .chain_mmr
+                .peaks_at(target_forest)
+                .expect("target_forest should be smaller than forest of the chain mmr");
+            let mut partial_mmr = PartialMmr::from_peaks(peaks);
+
+            for block_num in blocks.iter().map(BlockNumber::as_usize) {
+                // SAFETY: We have ensured block nums are less than chain length.
+                let leaf = state
+                    .chain_mmr
+                    .get(block_num)
+                    .expect("block num less than chain length should exist in chain mmr");
+                let path = state
+                    .chain_mmr
+                    .open_at(block_num, target_forest)
+                    .expect("block num and target forest should be valid for this mmr")
+                    .merkle_path;
```

Comment:
Not related to this PR, but I wonder if the `MmrProof` returned from `Mmr::open()` should also contain the leaf value. Let's create an issue for this in `miden-crypto`.

Reply:
Agreed, created https://github.com/0xPolygonMiden/crypto/issues/370.